core/lib/server/pub.ts

```diff
@@ -1337,48 +1327,63 @@ export async function getPubsWithRelatedValuesAndChildren<
 					.$if(Boolean(limit), (qb) => qb.limit(limit!))
 					.$if(Boolean(offset), (qb) => qb.offset(offset!))
 			)
-			.selectFrom("pub_tree")
+			.selectFrom("pub_tree as pt")
 			.select((eb) => [
-				"pub_tree.pubId",
-				"pub_tree.parentId",
-				"pub_tree.pubTypeId",
-				"pub_tree.depth",
-				"pub_tree.stageId",
-				"pub_tree.communityId",
-				"pub_tree.isCycle",
-				"pub_tree.path",
-				"pub_tree.createdAt",
-				"pub_tree.updatedAt",
-				"pub_tree.title",
-				jsonArrayFrom(
-					eb
-						.selectFrom("pub_tree as inner")
-						.select((eb) => [
-							"inner.valueId as id",
-							"inner.fieldId",
-							"inner.value",
-							"inner.relatedPubId",
-							"inner.valueCreatedAt as createdAt",
-							"inner.valueUpdatedAt as updatedAt",
-							"inner.schemaName",
-							"inner.slug as fieldSlug",
-							"inner.fieldName",
-						])
-						.whereRef("inner.pubId", "=", "pub_tree.pubId")
-						// this prevents us from double fetching values if we have detected a cycle
-						.whereRef("inner.depth", "=", "pub_tree.depth")
-						.where("inner.valueId", "is not", null)
-						.orderBy("inner.valueCreatedAt desc")
-				).as("values"),
+				"pt.pubId as id",
+				"pt.parentId",
+				"pt.pubTypeId",
+				"pt.depth",
+				"pt.stageId",
+				"pt.communityId",
+				"pt.isCycle",
+				"pt.path",
+				"pt.createdAt",
+				"pt.updatedAt",
+				"pt.title",
 			])
+			.$if(Boolean(withValues), (qb) =>
+				qb.select((eb) =>
+					jsonArrayFrom(
+						eb
+							.selectFrom("pub_values as pv")
+							.innerJoin("pub_fields", "pub_fields.id", "pv.fieldId")
+							.$if(Boolean(fieldSlugs), (qb) =>
+								qb.where("pub_fields.slug", "in", fieldSlugs!)
+							)
+							.select((eb) => [
+								"pv.id as id",
+								"pv.fieldId",
+								"pv.value",
+								"pv.relatedPubId",
+								"pv.createdAt as createdAt",
+								"pv.updatedAt as updatedAt",
+								"pub_fields.schemaName",
+								"pub_fields.slug as fieldSlug",
+								"pub_fields.name as fieldName",
+							])
+							.whereRef("pv.pubId", "=", "pt.pubId")
+							.orderBy("pv.createdAt desc")
+					).as("values")
+				)
+			)
```

Comment:
i think there's an even quicker way of doing this than this correlated subquery, but my brain is empty.
for now, this is efficient enough